src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs

```diff
@@ -400,19 +400,39 @@ public void OnCompleted(Func<object, Task> callback, object state)
 
         public void Flush()
         {
-            ProduceStartAndFireOnStarting().GetAwaiter().GetResult();
+            var produceStartTask = ProduceStartAndFireOnStarting();
+            if (produceStartTask.Status != TaskStatus.RanToCompletion)
+            {
+                produceStartTask.GetAwaiter().GetResult();
+            }
+
             SocketOutput.Write(_emptyData);
         }
 
-        public async Task FlushAsync(CancellationToken cancellationToken)
+        public Task FlushAsync(CancellationToken cancellationToken)
+        {
+            var produceStartTask = ProduceStartAndFireOnStarting();
+            if (produceStartTask.Status != TaskStatus.RanToCompletion)
+            {
+                return FlushAsyncAwaited(produceStartTask, cancellationToken);
+            }
+
+            return SocketOutput.WriteAsync(_emptyData, cancellationToken: cancellationToken);
+        }
+
+        public async Task FlushAsyncAwaited(Task produceStartTask, CancellationToken cancellationToken)
```

Comment:
`WriteAsync`+`WriteAsyncAwaited` are already there; however currently it will always call into `WriteAsyncAwaited`, construct a statemachine and `await` as the request hasn't started.

What it should be checking is if `ProduceStartAndFireOnStarting` produces an incompleted Task; if so then construct a statemachine and `await` it; otherwise just move onto the next step; which it will almost always do.